o Replace headers (Data Wrangling with Python pg. 154 – 163)

In [1]:
# import reader
from csv import reader

# read data files 
data_rdr = reader(open('DSC540-AllTrails data - nationalpark.csv', 'rb'))
header_rdr = reader(open('DSC540-nationalpark headers.csv', 'rb'))

# write object to new list
data_rows = [d for d in data_rdr]
# use list comprehension to remove mismatched headers
header_rows = [h for h in header_rdr if h[0] in data_rows[0]]

# make list of all short headers
all_short_headers = [h[0] for h in header_rows]

# initialize lists
skip_index = []
final_header_rows = []

# iterate over headers in dataset
for header in data_rows[0]:
    # single out headers that don't match
    if header not in all_short_headers:
        # return indexes we should skip
        index = data_rows[0].index(header)
        skip_index.append(index)
    else:
        for head in header_rows:
            if head[0] == header:
                final_header_rows.append(head)
                break
 

# create list for new data
new_data = []

# include only data rows
for row in data_rows[1:]:
    # initialize list
    new_row = []
    # use enumerate to isolate indexes of data rows to skip
    for i, d in enumerate(row):
        # make sure index is not in skip list
        if i not in skip_index:
            new_row.append(d)
    # add entry to new_data list
    new_data.append(new_row)

# initialize list
zipped_data = []

# iterate through new_data list
for drow in new_data:
    # zip each row and add it to new array
    zipped_data.append(zip(final_header_rows, drow))
    
print(zipped_data[0:1])


[[(['trail_id', 'Trail Identifier'], '10020048'), (['name', 'Trail Name'], 'Harding Ice Field Trail'), (['area_name', 'Park Name'], 'Kenai Fjords National Park'), (['city_name', 'City Name'], 'Seward'), (['state_name', 'State Name'], 'Alaska'), (['country_name', 'Country Name'], 'United States'), (['_geoloc', 'Geological Coordinates'], "{'lat': 60.18852, 'lng': -149.63156}"), (['popularity', 'Popularity Rating'], '24.8931'), (['length', 'Length of Hike'], '15610.598'), (['elevation_gain', 'Elevation Gain'], '1161.8976'), (['difficulty_rating', 'Difficulty Rating of Hike (1, 3, 5 or 7)'], '5'), (['route_type', 'Hiking Trail Route Type'], 'out and back'), (['visitor_usage', '# of Visitors'], '3'), (['avg_rating', 'Average Rating by Hikers (0-5 in .5 increments)'], '5'), (['num_reviews', 'Number of Reviews'], '423'), (['features', 'Trail Features'], "['dogs-no', 'forest', 'river', 'views', 'waterfall', 'wild-flowers', 'wildlife']"), (['activities', 'Park Activities'], "['birding', 'campin

In [2]:
# determine which fields to format
for x in enumerate(zipped_data[0]):
    print x


(0, (['trail_id', 'Trail Identifier'], '10020048'))
(1, (['name', 'Trail Name'], 'Harding Ice Field Trail'))
(2, (['area_name', 'Park Name'], 'Kenai Fjords National Park'))
(3, (['city_name', 'City Name'], 'Seward'))
(4, (['state_name', 'State Name'], 'Alaska'))
(5, (['country_name', 'Country Name'], 'United States'))
(6, (['_geoloc', 'Geological Coordinates'], "{'lat': 60.18852, 'lng': -149.63156}"))
(7, (['popularity', 'Popularity Rating'], '24.8931'))
(8, (['length', 'Length of Hike'], '15610.598'))
(9, (['elevation_gain', 'Elevation Gain'], '1161.8976'))
(10, (['difficulty_rating', 'Difficulty Rating of Hike (1, 3, 5 or 7)'], '5'))
(11, (['route_type', 'Hiking Trail Route Type'], 'out and back'))
(12, (['visitor_usage', '# of Visitors'], '3'))
(13, (['avg_rating', 'Average Rating by Hikers (0-5 in .5 increments)'], '5'))
(14, (['num_reviews', 'Number of Reviews'], '423'))
(15, (['features', 'Trail Features'], "['dogs-no', 'forest', 'river', 'views', 'waterfall', 'wild-flowers', 'wi

In [3]:
# format numeric fields to comma and 2 decimal places

# list of fields to format
field = [7, 8, 9]

# initialize dictionary to hold 
formatted_field = []


# iterate through rows
for row in zipped_data:
    # iterate through columns
    for col in field:
        # format to comma with two decimal places
        formatted_field = '{:,.2f}'.format(float(row[col][1]))
        # print trail_id, header, and newly formatted number
        print (row[0][1], row[col][0][1], formatted_field)



('10020048', 'Popularity Rating', '24.89')
('10020048', 'Length of Hike', '15,610.60')
('10020048', 'Elevation Gain', '1,161.90')
('10236086', 'Popularity Rating', '18.03')
('10236086', 'Length of Hike', '6,920.16')
('10236086', 'Elevation Gain', '507.80')
('10267857', 'Popularity Rating', '17.78')
('10267857', 'Length of Hike', '2,896.81')
('10267857', 'Elevation Gain', '81.99')
('10236076', 'Popularity Rating', '16.27')
('10236076', 'Length of Hike', '3,379.61')
('10236076', 'Elevation Gain', '119.79')
('10236082', 'Popularity Rating', '12.59')
('10236082', 'Length of Hike', '29,772.79')
('10236082', 'Elevation Gain', '1,124.71')
('10241620', 'Popularity Rating', '8.83')
('10241620', 'Length of Hike', '11,265.38')
('10241620', 'Elevation Gain', '1,178.97')
('10236080', 'Popularity Rating', '7.69')
('10236080', 'Length of Hike', '7,081.10')
('10236080', 'Elevation Gain', '287.73')
('10236075', 'Popularity Rating', '5.90')
('10236075', 'Length of Hike', '2,574.94')
('10236075', 'Elevat

('10269033', 'Popularity Rating', '2.44')
('10269033', 'Length of Hike', '66,465.74')
('10269033', 'Elevation Gain', '2,931.87')
('10348524', 'Popularity Rating', '2.43')
('10348524', 'Length of Hike', '11,748.18')
('10348524', 'Elevation Gain', '203.91')
('10268788', 'Popularity Rating', '2.38')
('10268788', 'Length of Hike', '49,567.67')
('10268788', 'Elevation Gain', '2,311.91')
('10268789', 'Popularity Rating', '2.37')
('10268789', 'Length of Hike', '36,371.08')
('10268789', 'Elevation Gain', '1,998.88')
('10259359', 'Popularity Rating', '2.37')
('10259359', 'Length of Hike', '8,046.70')
('10259359', 'Elevation Gain', '541.93')
('10272462', 'Popularity Rating', '2.36')
('10272462', 'Length of Hike', '7,081.10')
('10272462', 'Elevation Gain', '87.78')
('10261755', 'Popularity Rating', '2.33')
('10261755', 'Length of Hike', '4,345.22')
('10261755', 'Elevation Gain', '75.90')
('10394939', 'Popularity Rating', '2.26')
('10394939', 'Length of Hike', '8,207.63')
('10394939', 'Elevation G

('10027540', 'Elevation Gain', '423.98')
('10031434', 'Popularity Rating', '17.37')
('10031434', 'Length of Hike', '19,473.01')
('10031434', 'Elevation Gain', '365.76')
('10033216', 'Popularity Rating', '17.19')
('10033216', 'Length of Hike', '19,312.08')
('10033216', 'Elevation Gain', '1,012.85')
('10036262', 'Popularity Rating', '16.83')
('10036262', 'Length of Hike', '33,796.14')
('10036262', 'Elevation Gain', '1,349.96')
('10032041', 'Popularity Rating', '16.68')
('10032041', 'Length of Hike', '18,185.54')
('10032041', 'Elevation Gain', '775.72')
('10243572', 'Popularity Rating', '16.68')
('10243572', 'Length of Hike', '32,347.73')
('10243572', 'Elevation Gain', '370.94')
('10011461', 'Popularity Rating', '16.66')
('10011461', 'Length of Hike', '4,667.09')
('10011461', 'Elevation Gain', '174.96')
('10006763', 'Popularity Rating', '16.63')
('10006763', 'Length of Hike', '5,954.56')
('10006763', 'Elevation Gain', '257.86')
('10016234', 'Popularity Rating', '16.52')
('10016234', 'Leng

('10003960', 'Length of Hike', '9,977.91')
('10003960', 'Elevation Gain', '93.88')
('10037319', 'Popularity Rating', '7.86')
('10037319', 'Length of Hike', '20,760.49')
('10037319', 'Elevation Gain', '693.72')
('10031490', 'Popularity Rating', '7.86')
('10031490', 'Length of Hike', '114,102.21')
('10031490', 'Elevation Gain', '3,932.83')
('10284103', 'Popularity Rating', '7.85')
('10284103', 'Length of Hike', '6,598.29')
('10284103', 'Elevation Gain', '51.82')
('10029818', 'Popularity Rating', '7.83')
('10029818', 'Length of Hike', '11,426.31')
('10029818', 'Elevation Gain', '679.70')
('10273720', 'Popularity Rating', '7.82')
('10273720', 'Length of Hike', '34,439.88')
('10273720', 'Elevation Gain', '1,941.88')
('10028852', 'Popularity Rating', '7.79')
('10028852', 'Length of Hike', '5,310.82')
('10028852', 'Elevation Gain', '220.98')
('10267424', 'Popularity Rating', '7.78')
('10267424', 'Length of Hike', '6,759.23')
('10267424', 'Elevation Gain', '216.71')
('10299148', 'Popularity Ra

('10271978', 'Elevation Gain', '1,149.71')
('10261263', 'Popularity Rating', '6.27')
('10261263', 'Length of Hike', '7,402.96')
('10261263', 'Elevation Gain', '407.82')
('10009688', 'Popularity Rating', '6.26')
('10009688', 'Length of Hike', '38,463.23')
('10009688', 'Elevation Gain', '1,692.86')
('10262809', 'Popularity Rating', '6.26')
('10262809', 'Length of Hike', '78,535.79')
('10262809', 'Elevation Gain', '3,885.90')
('10264894', 'Popularity Rating', '6.25')
('10264894', 'Length of Hike', '42,486.58')
('10264894', 'Elevation Gain', '1,038.76')
('10263504', 'Popularity Rating', '6.23')
('10263504', 'Length of Hike', '9,656.04')
('10263504', 'Elevation Gain', '391.97')
('10032039', 'Popularity Rating', '6.13')
('10032039', 'Length of Hike', '4,506.15')
('10032039', 'Elevation Gain', '111.86')
('10024378', 'Popularity Rating', '6.12')
('10024378', 'Length of Hike', '482.80')
('10024378', 'Elevation Gain', '0.00')
('10028861', 'Popularity Rating', '6.12')
('10028861', 'Length of Hike

('10038390', 'Length of Hike', '32,186.80')
('10038390', 'Elevation Gain', '1,915.97')
('10266321', 'Popularity Rating', '2.36')
('10266321', 'Length of Hike', '482.80')
('10266321', 'Elevation Gain', '29.87')
('10041493', 'Popularity Rating', '2.35')
('10041493', 'Length of Hike', '2,253.08')
('10041493', 'Elevation Gain', '35.97')
('10264127', 'Popularity Rating', '2.34')
('10264127', 'Length of Hike', '8,690.44')
('10264127', 'Elevation Gain', '129.84')
('10262823', 'Popularity Rating', '2.33')
('10262823', 'Length of Hike', '46,188.06')
('10262823', 'Elevation Gain', '2,685.90')
('10312153', 'Popularity Rating', '2.33')
('10312153', 'Length of Hike', '33,313.34')
('10312153', 'Elevation Gain', '984.81')
('10002139', 'Popularity Rating', '2.32')
('10002139', 'Length of Hike', '26,393.18')
('10002139', 'Elevation Gain', '635.81')
('10420957', 'Popularity Rating', '2.30')
('10420957', 'Length of Hike', '142,426.59')
('10420957', 'Elevation Gain', '5,573.88')
('10267055', 'Popularity R

('10257657', 'Elevation Gain', '638.86')
('10331143', 'Popularity Rating', '1.54')
('10331143', 'Length of Hike', '40,233.50')
('10331143', 'Elevation Gain', '1,414.88')
('10310060', 'Popularity Rating', '1.50')
('10310060', 'Length of Hike', '47,475.53')
('10310060', 'Elevation Gain', '2,806.90')
('10369012', 'Popularity Rating', '1.46')
('10369012', 'Length of Hike', '66,626.68')
('10369012', 'Elevation Gain', '3,213.81')
('10480470', 'Popularity Rating', '1.45')
('10480470', 'Length of Hike', '32,347.73')
('10480470', 'Elevation Gain', '1,463.95')
('10327394', 'Popularity Rating', '1.45')
('10327394', 'Length of Hike', '91,249.58')
('10327394', 'Elevation Gain', '3,039.77')
('10003613', 'Popularity Rating', '1.44')
('10003613', 'Length of Hike', '5,310.82')
('10003613', 'Elevation Gain', '250.85')
('10290223', 'Popularity Rating', '1.44')
('10290223', 'Length of Hike', '84,812.22')
('10290223', 'Elevation Gain', '2,709.98')
('10259149', 'Popularity Rating', '1.41')
('10259149', 'Len

('10325911', 'Length of Hike', '3,540.55')
('10325911', 'Elevation Gain', '88.70')
('10293306', 'Popularity Rating', '3.71')
('10293306', 'Length of Hike', '20,116.75')
('10293306', 'Elevation Gain', '627.89')
('10258032', 'Popularity Rating', '3.54')
('10258032', 'Length of Hike', '1,609.34')
('10258032', 'Elevation Gain', '26.82')
('10503810', 'Popularity Rating', '3.37')
('10503810', 'Length of Hike', '12,874.72')
('10503810', 'Elevation Gain', '464.82')
('10327225', 'Popularity Rating', '3.35')
('10327225', 'Length of Hike', '14,805.93')
('10327225', 'Elevation Gain', '577.90')
('10328707', 'Popularity Rating', '3.31')
('10328707', 'Length of Hike', '12,552.85')
('10328707', 'Elevation Gain', '350.82')
('10506769', 'Popularity Rating', '3.17')
('10506769', 'Length of Hike', '16,254.33')
('10506769', 'Elevation Gain', '480.97')
('10236195', 'Popularity Rating', '3.15')
('10236195', 'Length of Hike', '6,598.29')
('10236195', 'Elevation Gain', '170.99')
('10301761', 'Popularity Rating

('10259207', 'Popularity Rating', '4.73')
('10259207', 'Length of Hike', '2,574.94')
('10259207', 'Elevation Gain', '36.88')
('10042890', 'Popularity Rating', '15.71')
('10042890', 'Length of Hike', '6,920.16')
('10042890', 'Elevation Gain', '64.92')
('10018166', 'Popularity Rating', '9.15')
('10018166', 'Length of Hike', '6,276.43')
('10018166', 'Elevation Gain', '50.90')
('10018164', 'Popularity Rating', '8.63')
('10018164', 'Length of Hike', '5,471.76')
('10018164', 'Elevation Gain', '78.94')
('10033779', 'Popularity Rating', '6.29')
('10033779', 'Length of Hike', '4,345.22')
('10033779', 'Elevation Gain', '43.89')
('10323594', 'Popularity Rating', '6.25')
('10323594', 'Length of Hike', '1,770.27')
('10323594', 'Elevation Gain', '28.96')
('10338879', 'Popularity Rating', '5.82')
('10338879', 'Length of Hike', '6,115.49')
('10338879', 'Elevation Gain', '40.84')
('10265729', 'Popularity Rating', '5.61')
('10265729', 'Length of Hike', '1,448.41')
('10265729', 'Elevation Gain', '26.82')

('10376356', 'Length of Hike', '22,852.63')
('10376356', 'Elevation Gain', '1,577.95')
('10036121', 'Popularity Rating', '13.90')
('10036121', 'Length of Hike', '13,196.59')
('10036121', 'Elevation Gain', '514.81')
('10235872', 'Popularity Rating', '13.56')
('10235872', 'Length of Hike', '4,184.28')
('10235872', 'Elevation Gain', '38.71')
('10240701', 'Popularity Rating', '13.53')
('10240701', 'Length of Hike', '5,632.69')
('10240701', 'Elevation Gain', '462.99')
('10235859', 'Popularity Rating', '13.52')
('10235859', 'Length of Hike', '16,093.40')
('10235859', 'Elevation Gain', '1,282.90')
('10032775', 'Popularity Rating', '13.39')
('10032775', 'Length of Hike', '1,931.21')
('10032775', 'Elevation Gain', '22.86')
('10376354', 'Popularity Rating', '12.72')
('10376354', 'Length of Hike', '18,346.48')
('10376354', 'Elevation Gain', '1,130.81')
('10249485', 'Popularity Rating', '12.49')
('10249485', 'Length of Hike', '3,057.75')
('10249485', 'Elevation Gain', '33.83')
('10294736', 'Popula

('10320598', 'Length of Hike', '9,173.24')
('10320598', 'Elevation Gain', '239.88')
('10342415', 'Popularity Rating', '4.68')
('10342415', 'Length of Hike', '18,668.34')
('10342415', 'Elevation Gain', '384.96')
('10344779', 'Popularity Rating', '4.36')
('10344779', 'Length of Hike', '11,104.45')
('10344779', 'Elevation Gain', '164.90')
('10012275', 'Popularity Rating', '4.28')
('10012275', 'Length of Hike', '1,126.54')
('10012275', 'Elevation Gain', '21.34')
('10348791', 'Popularity Rating', '4.02')
('10348791', 'Length of Hike', '3,540.55')
('10348791', 'Elevation Gain', '104.85')
('10294673', 'Popularity Rating', '3.78')
('10294673', 'Length of Hike', '2,896.81')
('10294673', 'Elevation Gain', '101.80')
('10344875', 'Popularity Rating', '3.50')
('10344875', 'Length of Hike', '9,173.24')
('10344875', 'Elevation Gain', '160.93')
('10235837', 'Popularity Rating', '3.37')
('10235837', 'Length of Hike', '1,287.47')
('10235837', 'Elevation Gain', '11.89')
('10406483', 'Popularity Rating', 

('10031102', 'Popularity Rating', '22.49')
('10031102', 'Length of Hike', '482.80')
('10031102', 'Elevation Gain', '32.92')
('10007994', 'Popularity Rating', '22.39')
('10007994', 'Length of Hike', '804.67')
('10007994', 'Elevation Gain', '4.88')
('10003094', 'Popularity Rating', '22.38')
('10003094', 'Length of Hike', '7,885.77')
('10003094', 'Elevation Gain', '217.93')
('10262194', 'Popularity Rating', '22.08')
('10262194', 'Length of Hike', '3,057.75')
('10262194', 'Elevation Gain', '136.86')
('10266145', 'Popularity Rating', '21.89')
('10266145', 'Length of Hike', '6,276.43')
('10266145', 'Elevation Gain', '197.82')
('10235759', 'Popularity Rating', '21.62')
('10235759', 'Length of Hike', '5,310.82')
('10235759', 'Elevation Gain', '44.81')
('10235914', 'Popularity Rating', '21.22')
('10235914', 'Length of Hike', '5,471.76')
('10235914', 'Elevation Gain', '251.76')
('10259638', 'Popularity Rating', '20.85')
('10259638', 'Length of Hike', '45,544.32')
('10259638', 'Elevation Gain', '

('10280405', 'Popularity Rating', '10.45')
('10280405', 'Length of Hike', '3,862.42')
('10280405', 'Elevation Gain', '175.87')
('10253568', 'Popularity Rating', '10.38')
('10253568', 'Length of Hike', '965.60')
('10253568', 'Elevation Gain', '12.80')
('10032799', 'Popularity Rating', '10.24')
('10032799', 'Length of Hike', '8,529.50')
('10032799', 'Elevation Gain', '203.00')
('10035613', 'Popularity Rating', '10.19')
('10035613', 'Length of Hike', '23,013.56')
('10035613', 'Elevation Gain', '545.90')
('10243229', 'Popularity Rating', '10.13')
('10243229', 'Length of Hike', '5,793.62')
('10243229', 'Elevation Gain', '324.92')
('10006672', 'Popularity Rating', '10.11')
('10006672', 'Length of Hike', '5,632.69')
('10006672', 'Elevation Gain', '236.83')
('10329449', 'Popularity Rating', '9.74')
('10329449', 'Length of Hike', '25,427.57')
('10329449', 'Elevation Gain', '568.76')
('10300060', 'Popularity Rating', '9.71')
('10300060', 'Length of Hike', '10,460.71')
('10300060', 'Elevation Gai

('10290795', 'Elevation Gain', '285.90')
('10268797', 'Popularity Rating', '4.98')
('10268797', 'Length of Hike', '9,495.11')
('10268797', 'Elevation Gain', '347.78')
('10299467', 'Popularity Rating', '4.93')
('10299467', 'Length of Hike', '9,012.30')
('10299467', 'Elevation Gain', '221.89')
('10002166', 'Popularity Rating', '4.87')
('10002166', 'Length of Hike', '9,656.04')
('10002166', 'Elevation Gain', '520.90')
('10301667', 'Popularity Rating', '4.84')
('10301667', 'Length of Hike', '10,460.71')
('10301667', 'Elevation Gain', '544.98')
('10034764', 'Popularity Rating', '4.82')
('10034764', 'Length of Hike', '1,287.47')
('10034764', 'Elevation Gain', '83.82')
('10321847', 'Popularity Rating', '4.64')
('10321847', 'Length of Hike', '965.60')
('10321847', 'Elevation Gain', '53.95')
('10282930', 'Popularity Rating', '4.63')
('10282930', 'Length of Hike', '482.80')
('10282930', 'Elevation Gain', '8.84')
('10356915', 'Popularity Rating', '4.54')
('10356915', 'Length of Hike', '9,012.30')

('10484068', 'Popularity Rating', '1.24')
('10484068', 'Length of Hike', '16,254.33')
('10484068', 'Elevation Gain', '783.95')
('10405344', 'Popularity Rating', '1.18')
('10405344', 'Length of Hike', '14,001.26')
('10405344', 'Elevation Gain', '719.94')
('10407786', 'Popularity Rating', '0.82')
('10407786', 'Length of Hike', '1,126.54')
('10407786', 'Elevation Gain', '3.96')
('10014805', 'Popularity Rating', '0.56')
('10014805', 'Length of Hike', '11,909.12')
('10014805', 'Elevation Gain', '659.89')
('10012756', 'Popularity Rating', '0.35')
('10012756', 'Length of Hike', '9,656.04')
('10012756', 'Elevation Gain', '648.92')
('10542445', 'Popularity Rating', '0.00')
('10542445', 'Length of Hike', '21,726.09')
('10542445', 'Elevation Gain', '965.91')
('10007491', 'Popularity Rating', '49.55')
('10007491', 'Length of Hike', '8,851.37')
('10007491', 'Elevation Gain', '463.91')
('10023187', 'Popularity Rating', '45.16')
('10023187', 'Length of Hike', '8,529.50')
('10023187', 'Elevation Gain'

('10272165', 'Elevation Gain', '1,536.80')
('10484495', 'Popularity Rating', '1.69')
('10484495', 'Length of Hike', '20,116.75')
('10484495', 'Elevation Gain', '762.91')
('10332059', 'Popularity Rating', '1.66')
('10332059', 'Length of Hike', '20,438.62')
('10332059', 'Elevation Gain', '981.76')
('10352623', 'Popularity Rating', '1.66')
('10352623', 'Length of Hike', '71,776.56')
('10352623', 'Elevation Gain', '4,494.89')
('10338708', 'Popularity Rating', '1.59')
('10338708', 'Length of Hike', '22,369.83')
('10338708', 'Elevation Gain', '722.99')
('10235757', 'Popularity Rating', '1.47')
('10235757', 'Length of Hike', '17,863.67')
('10235757', 'Elevation Gain', '255.73')
('10347402', 'Popularity Rating', '1.39')
('10347402', 'Length of Hike', '21,565.16')
('10347402', 'Elevation Gain', '1,347.83')
('10541288', 'Popularity Rating', '0.00')
('10541288', 'Length of Hike', '8,851.37')
('10541288', 'Elevation Gain', '64.92')
('10000719', 'Popularity Rating', '39.53')
('10000719', 'Length of

('10427569', 'Popularity Rating', '3.26')
('10427569', 'Length of Hike', '19,955.82')
('10427569', 'Elevation Gain', '719.94')
('10006161', 'Popularity Rating', '3.17')
('10006161', 'Length of Hike', '19,473.01')
('10006161', 'Elevation Gain', '368.81')
('10235923', 'Popularity Rating', '3.02')
('10235923', 'Length of Hike', '23,979.17')
('10235923', 'Elevation Gain', '471.83')
('10038551', 'Popularity Rating', '2.77')
('10038551', 'Length of Hike', '643.74')
('10038551', 'Elevation Gain', '11.89')
('10332755', 'Popularity Rating', '2.68')
('10332755', 'Length of Hike', '4,345.22')
('10332755', 'Elevation Gain', '146.00')
('10259492', 'Popularity Rating', '1.75')
('10259492', 'Length of Hike', '12,874.72')
('10259492', 'Elevation Gain', '404.77')
('10000825', 'Popularity Rating', '0.59')
('10000825', 'Length of Hike', '7,081.10')
('10000825', 'Elevation Gain', '123.75')
('10391957', 'Popularity Rating', '0.50')
('10391957', 'Length of Hike', '5,149.89')
('10391957', 'Elevation Gain', '

In [4]:
# identify bad data

# determine # of blanks
# create dictionary to keep track of blanks
blank_count ={}

# iterate through rows
for row in zipped_data:
    # iterate through columns
    for resp in row:
        header = resp[0][1]
        data = resp[1]
        # if length of data < 1 (blank)
        if len(data) < 1:
            # if header in list of blank keys increment counter
            if header in blank_count.keys():
                blank_count[header] += 1
            # if header not in list of blank keys then initialize counter
            else:
                blank_count[header] = 1
                
print 'Blanks:', blank_count



# determine # of zero values
# create dictionary to keep track of zeroes
zero_count ={}

# iterate through rows
for row in zipped_data:
    # iterate through columns
    for resp in row:
        header = resp[0][1]
        data = resp[1]
        # if field is zero
        if data == '0':
            # if header in list of zero keys increment counter
            if header in zero_count.keys():
                zero_count[header] += 1
            # if header not in list of zero keys initialize counter
            else:
                zero_count[header] = 1
                
print 'Zeroes:', zero_count



# determine # of NA's
# create dictionary to keep track of NAs
na_count ={}

# iterate through rows
for row in zipped_data:
# iterate through columsn
    for resp in row:
        header = resp[0][1]
        data = resp[1]
        # if field is 'NA'
        if data == 'NA':
            # if header in list of blank keys increment counter
            if header in zero_count.keys():
                na_count[header] += 1
            # if header not in list of blank keys initialize counter
            else:
                na_count[header] = 1
                
print "NA's:", na_count



Blanks: {'# of Visitors': 253}
Zeroes: {'Number of Reviews': 116, 'Elevation Gain': 25, 'Average Rating by Hikers (0-5 in .5 increments)': 116, 'Length of Hike': 2, 'Popularity Rating': 4}
NA's: {}


In [5]:
# identify outliers

# initialize dictionary
datatypes ={}

# setup start dictionary to ensure same type of data exists in each field
start_dict = {'digit': 0, 'boolean': 0, 'empty': 0, 'time_related': 0, 'text': 0, 'unknown': 0}


# iterate through rows
for row in zipped_data:
    # iterate through columns
    for resp in row:
        header = resp[0][1]
        data = resp[1]
        # set initial default key value to 'unknown'
        key = 'unknown'
        # determine whether string contains a digit 
        if data.isdigit():
            key = 'digit'
        # determine whether string contains a boolean value 
        elif data in ['Yes', 'No', 'True', 'False']:
            key = 'boolean'
        # determine whether string contains a space 
        elif data.isspace():
            key = 'empty'
        # determine whether string contains a '/' or ':' (i.e. time/date related data) 
        elif data.find('/') > 0 or data.find(':') < 0:
            key = 'time_related'
       # determine whether string contains an alpha character 
        elif data.isalpha():
            key = 'text'
        # determine whether header is in dictionary
        if header not in datatypes.keys():
            # create a separate dictionary object for each entry
            datatypes[header] = start_dict.copy()
            # increment key we found by 1
            datatypes[header][key] += 1                              
                                             
print datatypes


{'Elevation Gain': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, '# of Visitors': {'digit': 1, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 0, 'empty': 0}, 'Hiking Trail Route Type': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, 'City Name': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, 'Average Rating by Hikers (0-5 in .5 increments)': {'digit': 1, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 0, 'empty': 0}, 'State Name': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, 'Trail Name': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, 'Park Name': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, 'Country Name': {'digit': 0, 'boolean': 0, 'unknown': 0, 'text': 0, 'time_related': 1, 'empty': 0}, 'Popularity Rating': {'digit': 0, 'boolean': 0, 'unknown': 0

In [6]:
# import libraries
import numpy as np

# use numpy's unique method to return unique values
print np.unique(zipped_data, return_index=True)


(array([list(['_geoloc', 'Geological Coordinates']),
       list(['activities', 'Park Activities']),
       list(['area_name', 'Park Name']), ...,
       "{'lat': 63.80495, 'lng': -148.95416}",
       "{'lat': 63.80507, 'lng': -148.95336}",
       "{'lat': 63.88374, 'lng': -149.26764}"], dtype=object), array([ 12,  32,   4, ..., 697, 193, 481], dtype=int64))


In [7]:
# confirm that no duplicates for trail_id since it should be unique

# create a set containing the trail_ids
set_of_lines = set([x[0][1] for x in zipped_data])

# if the line number is unique, remove each key only once; if duplicates will return a key error
uniques = [x for x in zipped_data if not set_of_lines.remove(x[0][1])]

print set_of_lines
# empty set indicates no duplicates for trail_id


set([])


In [8]:
# check to see if duplicates for park name (as well as other fields)

# create a set containing the park name
set_of_lines = set([x[2][1] for x in zipped_data])

# if the line number is unique, remove each key only once; if duplicates will return a key error
uniques = [x for x in zipped_data if not set_of_lines.remove(x[2][1])]

print set_of_lines
# error raised since there are duplicates in all fields except for trail_id (separately tested each field)


KeyError: 'Kenai Fjords National Park'

Unable to install fuzzywuzzy - see error messages below:

(Python27) C:\Users\Tim>pip install fuzzywuzzy
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000000004EC17C8>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed',)': /simple/fuzzywuzzy/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000000004EC1F48>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed',)': /simple/fuzzywuzzy/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000000004EC16C8>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed',)': /simple/fuzzywuzzy/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000000004DE0D48>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed',)': /simple/fuzzywuzzy/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000000004E8B508>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed',)': /simple/fuzzywuzzy/
ERROR: Could not find a version that satisfies the requirement fuzzywuzzy (from versions: none)
ERROR: No matching distribution found for fuzzywuzzy

(Python27) C:\Users\Tim>